src/linalg.jl

```diff
@@ -343,28 +343,26 @@ function dot(A::AbstractSparseMatrixCSC{T1,S1},B::AbstractSparseMatrixCSC{T2,S2}
     return r
 end
 
-function dot(x::AbstractVector, A::AbstractSparseMatrixCSC, y::AbstractVector)
+function dot(x::AbstractVector{T1}, A::AbstractSparseMatrixCSC{T2}, y::AbstractVector{T3}) where {T1,T2,T3}
     require_one_based_indexing(x, y)
     m, n = size(A)
     (length(x) == m && n == length(y)) || throw(DimensionMismatch())
-    if iszero(m) || iszero(n)
-        return dot(zero(eltype(x)), zero(eltype(A)), zero(eltype(y)))
-    end
-    T = promote_type(eltype(x), eltype(A), eltype(y))
-    r = zero(T)
-    rvals = getrowval(A)
+    T = promote_type(T1, T2, T3)
+    s = zero(T)
```

Comment:
I'd suggest to make this "unit-safe" via

```suggestion
    s = dot(zero(T1), zero(T2), zero(T3))
    T = typeof(s)
```

Reply:
Is there any type where the second code is different from the original? Can the signature for multiplication be different from T*T returns T?

Reply:
Sure, for unitful numbers.

Reply:
To expand: if the units are different, then they can't even be promoted to some `T`. When they are all equal, they promote to themselves, but 3-arg dot with unit km each yields km^3, not km.